### PR TITLE
Add PHPUnit tests for Image Loading Optimization

### PR DESCRIPTION
**Pull request by westonruter**
## Summary

This is part of https://github.com/WordPress/performance/issues/869.

This adds full test coverage for the functions in the Image Loading Optimization module.

## Relevant technical choices

* The `ILO_HTML_Tag_Processor::open_tags()` generator is updated to ensure accurate breadcrumbs/XPath for SVG and MathML elements.
* The phpdoc and variable names are improved in `ilo_construct_preload_links()`.
* The `integrity` attribute is no longer copied from an `IMG` element to the preload `LINK`. This attribute doesn't appear on `IMG`, but rather is only on `LINK` and `SCRIPT`.
* Rework how timestamps are added to URL metrics by adding and examining them in `ilo_unshift_url_metrics()`. Previously the timestamp was added in `ilo_store_url_metric()` before passing to `ilo_construct_preload_links()`.
* Refactor `ilo_unshift_url_metrics()` to make it easier to test by passing more arguments.
* Fix JSON Schema definition in `ilo_register_endpoint()`.
* ~Discontinue excluding `WordPress-Extra` sniffs from applying to tests, and add specific sniff exclusions. By allowing `WordPress-Extra` to apply, PHPCBF can do automatic formatting. PHPCS errors were fixed in other test files.~ See #924.
* Fix REST API endpoint to reject storing a URL metric when the sample is already fully populated for the provided viewport. Previously it would continue to accept requests when _any_ of the breakpoints lacked a full sample size of URL metrics.
* Remove debug code from being returned in REST API response.

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

### CODE REVIEW

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (westonruter) [inline modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:202]: @dmsnell: FYI, I'm pretty happy with how `has_self_closing_flag` and the foreign element aspect of the HTML spec simplified things! Now it can properly get breadcrumbs/XPaths for each SVG/MathML element as well.

File: modules/images/image-loading-optimization/optimization.php
Comment (westonruter) [inline modules/images/image-loading-optimization/optimization.php:43]: The `empty( $lcp_element['attributes'] )` check is removed here because it is always set.

File: modules/images/image-loading-optimization/storage/rest-api.php
Comment (westonruter) [inline modules/images/image-loading-optimization/storage/rest-api.php]: I should address this.

File: composer.json
Comment (westonruter) [inline composer.json:31]: Added because `DOMDocument` is used in tests now.

File: modules/images/image-loading-optimization/optimization.php
Comment (westonruter) [inline modules/images/image-loading-optimization/optimization.php:175]: The `integrity` attribute doesn't exist on `IMG` so it shouldn't be copied.

File: tests/modules/images/image-loading-optimization/hooks-tests.php
Comment (westonruter) [inline tests/modules/images/image-loading-optimization/hooks-tests.php]:
```suggestion
class ILO_Hooks_Tests extends WP_UnitTestCase {
```

File: tests/modules/images/image-loading-optimization/detection-tests.php
Comment (westonruter) [inline tests/modules/images/image-loading-optimization/detection-tests.php]:
```suggestion
class ILO_Detection_Tests extends WP_UnitTestCase {
```

File: tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php
Comment (westonruter) [inline tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php]:
```suggestion
class ILO_HTML_Tag_Processor_Tests extends WP_UnitTestCase {
```

File: tests/modules/images/image-loading-optimization/optimization-tests.php
Comment (westonruter) [inline tests/modules/images/image-loading-optimization/optimization-tests.php]:
```suggestion
class ILO_Optimization_Tests extends WP_UnitTestCase {
```

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (mukeshpanchal27) [inline modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:246]: This file missing the `@since n.e.x.t` annotation for functions.

🔢 Apply to other functions also.

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (mukeshpanchal27) [inline modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:252]: Could we use `in_array` as it slightly more efficient as it could potentially exit earlier upon finding the first occurrence, while `foreach` iterates through the entire array.

```
if ( in_array( 'MATH', $this->open_stack_tags, true ) || in_array( 'SVG', $this->open_stack_tags, true ) ) {
	return true;
}
```

File: tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php]:
```suggestion
```
For test functions to execute correctly, we should maintain consistency by using either the `test_` prefix in function names or the `@test` annotation. In the mentioned PR, both methods were employed, so I suggest removing the `@test` annotation as it is unnecessary. This adjustment should also be applied to other tests for consistency. Thank you!

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (dmsnell) [reply modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:252]: @mukeshpanchal27 if you look closely you'll notice that this for loop bails as soon as it finds either a MATH or an SVG element. in contrast, with the dual `in_array()` checks we'd be iterating twice over the stack if neither tag is present whereas with the code as-written, we only iterate once.

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (westonruter) [reply modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:246]: Done in df0d09ea9872f28cf135e10277f4db14cba93501

File: tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php
Comment (westonruter) [reply tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php]: Done in 1fa877c7

File: modules/images/image-loading-optimization/optimization.php
Comment (felixarntz) [inline modules/images/image-loading-optimization/optimization.php:65]: Maybe a personal preference, but I find this notation overkill. This isn't used anywhere in WordPress. If we want to describe more complex shapes like this, we should probably introduce a class for them. This would be safer in any case than using an associative array which by definition takes any arbitrary data.

I think a simple `ILO_LCP_Element` class or similar would be helpful here. Then this would become `ILO_LCP_Element[]`.

Separate minor question: Why was this renamed to "LCP elements" instead of "LCP images"? Would it ever _not_ be an LCP image? If so, the choice of fields would be a bit odd, since almost all of them are specific to images.

File: modules/images/image-loading-optimization/optimization.php
Comment (felixarntz) [reply modules/images/image-loading-optimization/optimization.php:43]: Another benefit of using a class to represent LCP elements would be that that would be truly enforced :)

File: modules/images/image-loading-optimization/storage/data.php
Comment (felixarntz) [inline modules/images/image-loading-optimization/storage/data.php:119]: Not a blocker for this PR, but I think it would be useful to follow up once this is merged with another PR to introduce a class to represent a URL metric (e.g. `ILO_URL_Metric`). It would make the code much more intuitive and easier to follow as the fields would be enforced by the class definition.

File: modules/images/image-loading-optimization/storage/post-type.php
Comment (felixarntz) [inline modules/images/image-loading-optimization/storage/post-type.php:146]: This makes another good case why there should be a class to represent these objects. Validation would belong there, and probably be fewer lines of code as some of those checks would become irrelevant when enforcing proper fields via class properties.

File: modules/images/image-loading-optimization/storage/rest-api.php
Comment (felixarntz) [inline modules/images/image-loading-optimization/storage/rest-api.php:187]: Is the `ilo_needs_url_metric_for_breakpoint()` still valuable given it's no longer used here? Would it be possible to find a way to enhance the function in a way that it can be used here, rather than more inline logic in this function?

**Review by felixarntz — CHANGES_REQUESTED** [review]
This PR encompasses _a lot_ of changes, far more than only test coverage, which makes it difficult to review.

For now, I left a few comments on the production code changes. They mostly look good to me, although I think the code of some of those functions is extremely complex and could be simplified by separating responsibilities, and using smaller functions, or classes where applicable, for distinct processes. This is by no means a blocker for this PR, but I think it is critical to make this module/plugin long-term maintainable and more accessible to other contributors.

I think the PHPCS changes that affect the entire plugin make sense too, though I think that would be great to decouple as well, so that it can already be merged. FWIW, this probably shouldn't happen in the feature branch at all, as it likely will lead to more and more merge conflicts, the longer it takes for this feature branch to be merged. Could you open a PR with the PHPCS adjustments that don't affect the module directly against `trunk`?

For the production code changes and tests, I don't feel strongly that they need to be separated, though I believe that it could help getting the production code changes merged a lot faster. Reviewing the tests takes notably more time.

File: modules/images/image-loading-optimization/optimization.php
Comment (westonruter) [reply modules/images/image-loading-optimization/optimization.php:65]: > Maybe a personal preference, but I find this notation overkill. This isn't used anywhere in WordPress. If we want to describe more complex shapes like this, we should probably introduce a class for them. This would be safer in any case than using an associative array which by definition takes any arbitrary data.

Actually, it is used in WordPress, specifically [in the `sodium_compat` code](https://github.com/search?q=repo%3AWordPress%2Fwordpress-develop%20%2Farray%7B%2F&type=code). Granted, this is not really organic WordPress code. In any case, with precise phpdoc typing here, PHPStan will detect any anomalies with static analysis. So it's trying to gain the same benefits as TypeScript gives.

> I think a simple `ILO_LCP_Element` class or similar would be helpful here. Then this would become `ILO_LCP_Element[]`.

Yeah, I like that idea. However, believe it or not, I was wanting to limit the refactor of the existing code in this PR once the test coverage was completed. With the coverage in place, refactoring will be safer as I'll be able to detect regressions better.

> Separate minor question: Why was this renamed to "LCP elements" instead of "LCP images"? Would it ever not be an LCP image? If so, the choice of fields would be a bit odd, since almost all of them are specific to images.

Yes, for example in https://github.com/WordPress/performance/pull/914 this is extended to support elements with CSS background images.

File: modules/images/image-loading-optimization/storage/post-type.php
Comment (westonruter) [reply modules/images/image-loading-optimization/storage/post-type.php:146]: Nevertheless, the data would be stored as JSON in the database and not as a serialized class. So this validation would still have to be done somewhere.

File: modules/images/image-loading-optimization/optimization.php
Comment (felixarntz) [reply modules/images/image-loading-optimization/optimization.php:65]: > Yeah, I like that idea. However, believe it or not, I was wanting to limit the refactor of the existing code in this PR once the test coverage was completed. With the coverage in place, refactoring will be safer as I'll be able to detect regressions better.

+1, that makes total sense to me. Maybe open an issue already?

File: modules/images/image-loading-optimization/storage/post-type.php
Comment (felixarntz) [reply modules/images/image-loading-optimization/storage/post-type.php:146]: Right, I think it could then be done in the class. I would expect a class to e.g. receive this associative array as constructor parameter and then validate it.

File: modules/images/image-loading-optimization/storage/rest-api.php
Comment (westonruter) [reply modules/images/image-loading-optimization/storage/rest-api.php:187]: Good point! Yeah, it's not valuable. I've removed it in 3da1363c. I did try to consolidate this logic into that existing function, but the two codepaths didn't mesh well.

File: modules/images/image-loading-optimization/optimization.php
Comment (westonruter) [inline modules/images/image-loading-optimization/optimization.php:148]: Some cool things with PHPStan's static analysis here. If 0 is supplied as the second arg to `array_column()` it detects an error:

> phpstan: Call to function in_array() with arguments true, list<int> and true will always evaluate to false.

This is because from the array shape returned by `ilo_get_needed_minimum_viewport_widths()` it knows that the first column will all be integers (the minimum viewport widths) and thus the `in_array()` check will always be `false`. 

Conversely, if I supply a higher index like 2 in `array_column()`, I get:

> phpstan: Call to function in_array() with arguments true, array{} and true will always evaluate to false.

Here PHPStan knows that since there are only two items in each tuple, the return of `array_column()` will be an empty array, and thus `in_array()` will also always return `false`.

**Comment by westonruter** [comment]
@felixarntz 

> For now, I left a few comments on the production code changes. They mostly look good to me, although I think the code of some of those functions is extremely complex and could be simplified by separating responsibilities, and using smaller functions, or classes where applicable, for distinct processes. This is by no means a blocker for this PR, but I think it is critical to make this module/plugin long-term maintainable and more accessible to other contributors. [...] For the production code changes and tests, I don't feel strongly that they need to be separated, though I believe that it could help getting the production code changes merged a lot faster. Reviewing the tests takes notably more time.

Agreed. As mentioned above, I wanted to limit further refactoring until tests were written. The code changes not related to test coverage were primarily to fix blatant issues I uncovered while writing the tests. Sorry it grew so large. Since the review cycles have been pretty long, it's hard to keep the PRs very concise.

> I think the PHPCS changes that affect the entire plugin make sense too, though I think that would be great to decouple as well, so that it can already be merged. FWIW, this probably shouldn't happen in the feature branch at all, as it likely will lead to more and more merge conflicts, the longer it takes for this feature branch to be merged. Could you open a PR with the PHPCS adjustments that don't affect the module directly against `trunk`?

Agreed. Done in https://github.com/WordPress/performance/pull/924. I've reverted the changes in this PR via 2181d884305abce6a95182fc8b037beed747bd9b.

**Comment by westonruter** [comment]
Unit tests are failing due to `Audit_Enqueued_Assets_Tests`, which isn't related to the changes in this PR. ~Two of the failures are fixed in `trunk` and~ the other is fixed in #926 (which for some reason doesn't happen in `trunk`).

**Comment by westonruter** [comment]
The PHPStan failures are likewise not related to code changes here. They may be resolved by merging `trunk` into the feature branch, but I will wait to do that until #926 is merged.

File: modules/images/image-loading-optimization/optimization.php
Comment (westonruter) [reply modules/images/image-loading-optimization/optimization.php:65]: I've just now added it as a todo item on the [overview issue](https://github.com/WordPress/performance/issues/869).

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (felixarntz) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]: I was just wondering why this function is in the `storage/data.php` file. It's only called in `optimization.php`, and the function name to me suggests it would fit better in there.

I then noticed `storage/data.php` contains several other functions like this, that to me don't seem storage related. Is that intentional? If so, can you clarify how you see the responsibilities between these two files?

Maybe there's a more intuitive way to break apart the functions here. To be fair, some of it may sort of automatically be addressed when using classes for some of those things as applicable.

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (felixarntz) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]: Nit-pick: I find the return of 0, 1, or 2 an odd choice. I realize `wp_verify_nonce()` returns that, but it's extremely unintuitive, and I don't think many people know, except from reading the docs.

There's no value for us (and probably any other use-cases) to make a difference between when the nonce was generated, so I'd prefer to just return a boolean (i.e. cast the result to a bool). You aren't testing the `2` return value here either as far as I can tell (which makes sense IMO!), so that confirms it's useless to us :)

**Review by felixarntz — APPROVED** [review]
Thanks @westonruter, LGTM! I think this is good to merge once the PHPStan problems have been addressed (here or via the other PR).

Let's separately follow up on simplifying some of the logic with classes, and also file structure (see my comment below).

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (westonruter) [reply tests/modules/images/image-loading-optimization/storage/data-tests.php]: Yeah, sounds good. Done in d865bacb.

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (westonruter) [reply tests/modules/images/image-loading-optimization/storage/data-tests.php]: Good catch! Function moved in ed305a9efeba2411f37cd1eaf69f7258d2828fb6.

The other functions are generally storage related in that they relate to how fresh the stored data is, how to determine the slug for the data, manipulating the stored data, etc. But I agree this the functions in this file are much more of a grab bag compared to the functions in other files. I'll work to organize them better with the refactoring.

File: modules/images/image-loading-optimization/class-ilo-html-tag-processor.php
Comment (mukeshpanchal27) [inline modules/images/image-loading-optimization/class-ilo-html-tag-processor.php:246]:
```suggestion
	 *
	 * @since n.e.x.t
	 *
```

File: tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php:287]: Nice if we add `@param` that used in `dataProvider`.

File: tests/modules/images/image-loading-optimization/detection-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/detection-tests.php]:
```suggestion
	 * @covers ::ilo_get_detection_script
	 *
	 * @dataProvider data_provider_ilo_get_detection_script
```

File: tests/modules/images/image-loading-optimization/optimization-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/optimization-tests.php]:
```suggestion
	 * @covers ::ilo_can_optimize_response
	 *
	 * @dataProvider data_provider_test_ilo_can_optimize_response
```
Nice if we add @param that used in dataProvider.

File: tests/modules/images/image-loading-optimization/optimization-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/optimization-tests.php]:
```suggestion
	 * @covers ::ilo_construct_preload_links
	 *
	 * @dataProvider data_provider_test_ilo_construct_preload_links
```

File: tests/modules/images/image-loading-optimization/optimization-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/optimization-tests.php]:
```suggestion
	 * @covers ::ilo_optimize_template_output_buffer
	 *
	 * @dataProvider data_provider_test_ilo_optimize_template_output_buffer
```

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]:
```suggestion
	 * @covers ::ilo_get_normalized_query_vars
	 *
	 * @dataProvider data_provider_test_ilo_get_normalized_query_vars
```

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]:
```suggestion
	 * @covers ::ilo_unshift_url_metrics
	 *
	 * @dataProvider data_provider_sample_size_and_breakpoints
```

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]:
```suggestion
	 * @covers ::ilo_group_url_metrics_by_breakpoint
	 *
	 * @dataProvider data_provider_test_ilo_group_url_metrics_by_breakpoint
```

File: tests/modules/images/image-loading-optimization/storage/data-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/data-tests.php]:
```suggestion
	 * @covers ::ilo_get_needed_minimum_viewport_widths
	 *
	 * @dataProvider data_provider_test_ilo_get_needed_minimum_viewport_widths
```

File: tests/modules/images/image-loading-optimization/storage/lock-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/lock-tests.php]:
```suggestion
	 * @covers ::ilo_get_url_metric_storage_lock_ttl
	 *
	 * @dataProvider data_provider_ilo_get_url_metric_storage_lock_ttl
```

File: tests/modules/images/image-loading-optimization/storage/post-type-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/post-type-tests.php]:
```suggestion
	 * @covers ::ilo_parse_stored_url_metrics
	 *
	 * @dataProvider data_provider_test_ilo_parse_stored_url_metrics
```

File: tests/modules/images/image-loading-optimization/storage/rest-api-tests.php
Comment (mukeshpanchal27) [inline tests/modules/images/image-loading-optimization/storage/rest-api-tests.php]:
```suggestion
	 * @covers ::ilo_handle_rest_request
	 *
	 * @dataProvider data_provider_invalid_params
```

**Review by mukeshpanchal27 — APPROVED** [review]
Thank you, @westonruter, for the updates. I've left some minor suggestions, but they don't pose any blockers.

- It would be nice to add `@param` annotations for the parameters used in the `dataProvider` for tests.

File: tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php
Comment (westonruter) [reply tests/modules/images/image-loading-optimization/class-ilo-html-tag-processor-tests.php:287]: Yeah, somewhat. But it looks like most of the other tests using data providers don't have `@param` tags in the plugin. Not that this is a good reason.

**Pull request merged**
